heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerMain.java

```diff
@@ -347,14 +347,18 @@ protected boolean validateRuntimeManage(
       return false;
     }
 
-    // Check whether cluster/role/environ matched
-    ExecutionEnvironment.ExecutionState executionState = adaptor.getExecutionState(topologyName);
-    if (executionState == null
-        || !executionState.getCluster().equals(Context.cluster(config))
-        || !executionState.getRole().equals(Context.role(config))
-        || !executionState.getEnviron().equals(Context.environ(config))) {
-      LOG.severe("cluster/role/environ not matched");
-      return false;
+    // Check whether cluster/role/environ matched. It is possible to get into a state where some of
+    // the data does not actually exist in zookeeper. If we are in this state them assume the
+    // runtime is valid.
+    if (adaptor.doesExecutionStateExist(topologyName)) {
```

Comment:
Looking closer at the code I think this method would be better named ` validatePrecommandState` or something along those lines. It's basically asserting that the command is ok to be executed based on alignment of client state with zookeeper. If executionState doesn't exist in ZK most commands probably should not be run (active, deactivate, restart). If we want to permit `kill` to run, then I suggest we change this method to also take the command and use that is an input.

cc/ @maosongfu for his thoughts as well.